File: farmacies.py
import tkinter as tk
from tkinter import ttk
import requests

def get_data():
    response = requests.get('https://pompeu.neocities.org/farmacies/farmacies.json')
    if response.status_code != 200:
        raise ValueError(f'HTTP error! status: {response.status_code}')
    else:
        return response.json()

def search():
    postal_code = entry.get()
    data = get_data()
    search_results = [pharmacy for pharmacy in data if pharmacy['CODI POSTAL'] == postal_code]
    for i, result in enumerate(search_results):
        table.insert('', 'end', values=(i+1, result['NOM FARMACIA'], result['CODI POSTAL'], result['CODI ABS'], result['NOM ABS'], result['TIPUS VIA'], result['NOM VIA'], result['NUMERO VIA'], result['NOM MUNICIPI']))

root = tk.Tk()
root.title("Pharmacy Search")

#title 
title = tk.Label(root, text="Pharmacy Search", font=("Arial", 16))
title.grid(row=0, column=0, columnspan=3, padx=5, pady=5, ipadx=5, ipady=5)

#explanation
explanation = tk.Label(root, text="This is a Search to find Catalan pharmacies from their postal code")
explanation.grid(row=1, column=0, columnspan=3, padx=5, pady=5, ipadx=5, ipady=5)

#label
label = tk.Label(root, text="Postal Code:")
label.grid(row=2, column=0, padx=5, pady=5)

#entry
entry = tk.Entry(root)
entry.grid(row=2, column=1, padx=5, pady=5)

#search button
search_button = tk.Button(root, text="Search", command=search)
search_button.grid(row=2, column=2, padx=5, pady=5)

#table
table = ttk.Treeview(root, columns=("No", "Pharmacy Name", "Postal Code", "ABS Code", "ABS Name", "Type of Street", "Street Name", "Street Number", "Municipality"), show="headings")
table.heading("No", text="No.")
table.heading("Pharmacy Name", text="Pharmacy Name")
table.heading("Postal Code", text="Postal Code")
table.heading("ABS Code", text="ABS Code")
table.heading("ABS Name", text="ABS Name")
table.heading("Type of Street", text="Type of Street")
table.heading("Street Name", text="Street Name")
table.heading("Street Number", text="Street Number")
table.heading("Municipality", text="Municipality")
table.column("No", width=50)
table.column("Pharmacy Name", width=150)
table.column("Postal Code", width=100)
table.column("ABS Code", width=100)
table.column("ABS Name", width=150)
table.column("Type of Street", width=150)
table.column("Street Name", width=150)
table.column("Street Number", width=100)
table.column("Municipality", width=150)
table.grid(row=3, column=0, columnspan=3, padx=5, pady=5)

root.mainloop()

File: farmacies.cpp
#include <QApplication>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QJsonDocument>
#include <QJsonArray>
#include <QTableWidget>
#include <QVBoxLayout>
#include <QLineEdit>
#include <QPushButton>
#include <QLabel>
#include <QHeaderView>

class PharmacySearch : public QWidget {

public:
    PharmacySearch(QWidget *parent = nullptr) : QWidget(parent) {
        setWindowTitle("Pharmacy Search");

        // Title
        auto title = new QLabel("Pharmacy Search", this);
        title->setStyleSheet("font: 16pt Arial;");
        title->setAlignment(Qt::AlignCenter);

        // Explanation
        auto explanation = new QLabel("This is a Search to find Catalan pharmacies from their postal code", this);
        explanation->setAlignment(Qt::AlignCenter);

        // label
        auto label = new QLabel("Postal Code:", this);

        // entry
        entry = new QLineEdit(this);

        // search button
        auto searchButton = new QPushButton("Search", this);
        connect(searchButton, &QPushButton::clicked, this, &PharmacySearch::search);

        // table
        table = new QTableWidget(this);
        table->setColumnCount(9);
        table->setHorizontalHeaderLabels({"No", "Pharmacy Name", "Postal Code", "ABS Code", "ABS Name", "Type of Street", "Street Name", "Street Number", "Municipality"});
        table->horizontalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);
        table->setEditTriggers(QAbstractItemView::NoEditTriggers);

        // layout
        auto mainLayout = new QVBoxLayout(this);
        mainLayout->addWidget(titlemainLayout->addWidget(explanation);
        mainLayout->addWidget(label);
        mainLayout->addWidget(entry);
        mainLayout->addWidget(searchButton);
        mainLayout->addWidget(table);

        setLayout(mainLayout);
    }

private:
    void search() {
        QString postalCode = entry->text();
        QNetworkAccessManager *manager = new QNetworkAccessManager(this);
        connect(manager, &QNetworkAccessManager::finished, this, [=](QNetworkReply *reply) {
            if (reply->error()) {
                throw std::runtime_error(QString("Error while fetching data: %1").arg(reply->errorString()).toStdString());
            } else {
                QJsonDocument document = QJsonDocument::fromJson(reply->readAll());
                QJsonArray array = document.array();
                for (int i = 0; i < array.size(); ++i) {
                    QJsonObject obj = array.at(i).toObject();
                    if (obj["CODI POSTAL"].toString() == postalCode) {
                        int row = table->rowCount();
                        table->insertRow(row);
                        table->setItem(row, 0, new QTableWidgetItem(QString::number(row + 1)));
                        table->setItem(row, 1, new QTableWidgetItem(obj["NOM FARMACIA"].toString()));
                        table->setItem(row, 2, new QTableWidgetItem(obj["CODI POSTAL"].toString()));
                        table->setItem(row, 3, new QTableWidgetItem(obj["CODI ABS"].toString()));
                        table->setItem(row, 4, new QTableWidgetItem(obj["NOM ABS"].toString()));
                        table->setItem(row, 5, new QTableWidgetItem(obj["TIPUS VIA"].toString()));
                        table->setItem(row, 6, new QTableWidgetItem(obj["NOM VIA"].toString()));
                        table->setItem(row, 7, new QTableWidgetItem(obj["NUMERO VIA"].toString()));
                        table->setItem(row, 8, new QTableWidgetItem(obj["NOM MUNICIPI"].toString()));
                    }
                }
        });
        manager->get(QNetworkRequest(QUrl("https://pompeu.neocities.org/farmacies/farmacies.json")));
    }

    QLineEdit *entry;
    QTableWidget *table;
};

int main(int argc, char *argv[]) {
    QApplication a(argc, argv);
    PharmacySearch w;
    w.show();
    return a.exec();
}

